pandas/_libs/lib.pyx

```diff
@@ -669,6 +669,40 @@ def astype_intsafe(ndarray[object] arr, cnp.dtype new_dtype) -> ndarray:
 
     return result
 
+ctypedef fused ndarr_object:
+    ndarray[object, ndim=1]
+    ndarray[object, ndim=2]
+
+# TODO: get rid of this in StringArray and modify
+#  and go through ensure_string_array instead
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def convert_nans_to_NA(ndarr_object arr) -> ndarray:
+    """
+    Helper for StringArray that converts null values that
+    are not pd.NA(e.g. np.nan, None) to pd.NA. Assumes elements
+    have already been validated as null.
+    """
+    cdef:
+        Py_ssize_t i, m, n
+        object val
+        ndarr_object result
+    result = np.asarray(arr, dtype="object")
```

Comment:
Hm. I probably can operate inplace on arr here, if the constructor handles making copies.

Also, interesting question: What would happen if ``copy=False`` was passed to the StringArray constructor and the data contained nans? Would the nans of the underlying array the StringArray was converted from be converted to pd.NA?

Reply:
i think this `result` is going to be exactly `arr`, so this is de-facto operating in-place.